src/code/z_skin_matrix.c

```diff
@@ -1,42 +1,728 @@
+    s32 this_col;
+    u32 pad;
+
+    SkinMatrix_MtxFCopy(src, &mfCopy);
+    SkinMatrix_Clear(dest);
+    for (this_row = 0; this_row != 4; this_row++) {
+        for (this_col = this_row; (this_col < 4) && (fabsf(mfCopy.mf[this_row][this_col]) < 0.0005f); this_col++) {}
+        if (this_col == 4) {
+            // reaching col = 4 means the row is either all 0 or a duplicate row.
+            // therefore singular matrix (0 determinant).
+            VT_COL(YELLOW, BLACK);
+            osSyncPrintf("Skin_Matrix_InverseMatrix():逆行列つくれません\n");
+            osSyncPrintf(VT_RST);
+            return 2;
+        }
+
+        if (this_col != this_row) { // responsible for swapping columns if zero on diagonal
+            for (i = 0; i != 4; i++) {
+                temp1 = mfCopy.mf[i][this_col];
+                mfCopy.mf[i][this_col] = mfCopy.mf[i][this_row];
+                mfCopy.mf[i][this_row] = temp1;
+
+                temp1 = dest->mf[i][this_col];
+                dest->mf[i][this_col] = dest->mf[i][this_row];
+                dest->mf[i][this_row] = temp1;
+            }
+        }
+
+        diag_element = mfCopy.mf[this_row][this_row];
+        // Scale this whole column s.t. the diag element = 1
+        mfCopy.mf[0][this_row] = mfCopy.mf[0][this_row] / diag_element;
+        dest->mf[0][this_row] = dest->mf[0][this_row] / diag_element;
+        mfCopy.mf[1][this_row] = mfCopy.mf[1][this_row] / diag_element;
+        dest->mf[1][this_row] = dest->mf[1][this_row] / diag_element;
+        mfCopy.mf[2][this_row] = mfCopy.mf[2][this_row] / diag_element;
+        dest->mf[2][this_row] = dest->mf[2][this_row] / diag_element;
+        mfCopy.mf[3][this_row] = mfCopy.mf[3][this_row] / diag_element;
+        dest->mf[3][this_row] = dest->mf[3][this_row] / diag_element;
+
+        // col i = col i - a * col j
+        for (k = 0; k != 4; k++) {
+            if (k != this_row) {
+                temp2 = mfCopy.mf[this_row][k];
+                mfCopy.mf[0][k] = mfCopy.mf[0][k] - (mfCopy.mf[0][this_row] * temp2);
+                dest->mf[0][k] = dest->mf[0][k] - (dest->mf[0][this_row] * temp2);
+                mfCopy.mf[1][k] = mfCopy.mf[1][k] - (mfCopy.mf[1][this_row] * temp2);
+                dest->mf[1][k] = dest->mf[1][k] - (dest->mf[1][this_row] * temp2);
+                mfCopy.mf[2][k] = mfCopy.mf[2][k] - (mfCopy.mf[2][this_row] * temp2);
+                dest->mf[2][k] = dest->mf[2][k] - (dest->mf[2][this_row] * temp2);
+                mfCopy.mf[3][k] = mfCopy.mf[3][k] - (mfCopy.mf[3][this_row] * temp2);
+                dest->mf[3][k] = dest->mf[3][k] - (dest->mf[3][this_row] * temp2);
+            }
+        }
+    }
+    return 0;
+}
+#else
+#pragma GLOBAL_ASM("asm/non_matchings/code/z_skin_matrix/SkinMatrix_Invert.s")
+#endif
+
+/**
+ * Produces a matrix which scales x,y,z components of vectors or x,y,z rows of matrices (when applied on LHS)
+ */
+void SkinMatrix_SetScaling(MtxF* mf, f32 x, f32 y, f32 z) {
+    mf->xy = 0.0f;
```

Comment:
the corresponding function in sys_matrix.c is simply called Matrix_Scale, is there any reason it needs to say scaling in this context?

Reply:
i guess they arent really the same function, but im still wondering is Scale or SetScale would be better

Reply:
I renamed it this due to Roman's suggestion (naming things "SetX" to clarify that these set an existing matrix rather than creating a new one like in sys_matrix).

Tha name "scaling" is the name of this type of matrix in the literature, I forgot that sys_matrix has a weird name for it lol. I'll set it back to "Scale" for now to be consistent with the codebase but would prefer if we changed all the matrix names to be consistent with the actual mathematical names for these objects

Reply:
I see, yeah I just wasnt sure if thats the "proper" way to call it, as I know next to nothing about matrix math.
If anything, now is the time to change them in sys_matrix and make everything proper and consistent, but interested what others think.